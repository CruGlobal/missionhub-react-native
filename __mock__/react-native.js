import { NativeModules } from 'react-native';

// From https://github.com/facebook/react-native/issues/18279#issuecomment-374177940
// Will be made unnecessary once https://github.com/facebook/react-native/pull/18718 is released. It doesn't seem to be in 0.55.3
NativeModules.BlobModule = {
  ...NativeModules.BlobModule,
  addNetworkingHandler: jest.fn(),
};
