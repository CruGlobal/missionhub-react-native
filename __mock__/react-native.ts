jest.mock(
  'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
  () => ({
    // @ts-ignore
    ...jest.requireActual(
      'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
    ),
    getConstants: () => ({
      forceTouchAvailable: false,
      interfaceIdiom: 'en',
      isTesting: false,
      osVersion: 'ios',
      reactNativeVersion: { major: 60, minor: 1, patch: 0 },
      systemName: 'ios',
    }),
  }),
);
