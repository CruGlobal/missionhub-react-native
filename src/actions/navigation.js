import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
