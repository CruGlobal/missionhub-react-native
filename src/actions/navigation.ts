/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
