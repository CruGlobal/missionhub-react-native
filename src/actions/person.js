import callApi, { REQUESTS } from './api';
import { PERSON_FIRST_NAME_CHANGED, PERSON_LAST_NAME_CHANGED } from '../constants';


export function personFirstNameChanged(firstName) {
  return {
    type: PERSON_FIRST_NAME_CHANGED,
    personFirstName: firstName,
  };
}

export function personLastNameChanged(lastName) {
  return {
    type: PERSON_LAST_NAME_CHANGED,
    personLastName: lastName,
  };
}

export function saveNotes(personId, notes, firstSave) {
  return (dispatch, getState) => {
    const myId = getState().auth.personId;
    if (!personId || !notes) {
      return Promise.reject('InvalidData', personId, notes);
    }

    const bodyData = {
      data: {
        type: 'person_note',
        attributes: {
          content: notes,
        },
      },
      relationships: {
        person: {
          data: {
            type: 'person',
            id: personId,
          },
        },
      },
      user: {
        data: {
          type: 'user',
          id: myId,
        },
      },
    };

    const query = {};

    if (firstSave) return dispatch(callApi(REQUESTS.ADD_PERSON_NOTES, query, bodyData));
    return dispatch(callApi(REQUESTS.UPDATE_PERSON_NOTES, query, bodyData));
  };
}