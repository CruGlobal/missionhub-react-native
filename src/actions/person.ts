/* eslint complexity: 0, max-lines: 0, max-lines-per-function: 0, max-params: 0 */

import {
  CONTACT_PERSON_SCREEN,
  IS_USER_CREATED_MEMBER_PERSON_SCREEN,
  IS_GROUPS_MEMBER_PERSON_SCREEN,
  MEMBER_PERSON_SCREEN,
  ME_PERSONAL_PERSON_SCREEN,
  IS_GROUPS_ME_COMMUNITY_PERSON_SCREEN,
  ME_COMMUNITY_PERSON_SCREEN,
} from '../containers/Groups/AssignedPersonScreen/';
import { UNASSIGNED_PERSON_SCREEN } from '../containers/Groups/UnassignedPersonScreen';
import {
  UPDATE_PERSON_ATTRIBUTES,
  DELETE_PERSON,
  ACTIONS,
  LOAD_PERSON_DETAILS,
  ORG_PERMISSIONS,
} from '../constants';
import { hasOrgPermissions, exists } from '../utils/common';
import {
  personSelector,
  orgPermissionSelector,
  contactAssignmentSelector,
} from '../selectors/people';
import { organizationSelector } from '../selectors/organizations';
import { REQUESTS } from '../api/routes';

import callApi from './api';
import { trackActionWithoutData } from './analytics';
import { navigatePush } from './navigation';
import { getMyCommunities } from './organizations';
import { getMySteps } from './steps';

// @ts-ignore
export function getMe(extraInclude) {
  const personInclude =
    'email_addresses,phone_numbers,organizational_permissions.organization,reverse_contact_assignments,user';

  const include = extraInclude
    ? `${personInclude},${extraInclude}`
    : personInclude;

  // @ts-ignore
  return async dispatch => {
    const { response: person } = await dispatch(
      callApi(REQUESTS.GET_ME, { include }),
    );
    return person;
  };
}

// @ts-ignore
export function getPersonDetails(id, orgId) {
  const personInclude =
    'contact_assignments.person,email_addresses,phone_numbers,organizational_permissions.organization,reverse_contact_assignments,user';

  // @ts-ignore
  return async dispatch => {
    if (!id) {
      return Promise.reject(
        'Invalid Data from getPersonDetails: no personId passed in',
      );
    }

    const query = {
      person_id: id,
      include: personInclude,
    };
    const { response: person } = await dispatch(
      callApi(REQUESTS.GET_PERSON, query),
    );
    const orgPermission =
      orgId &&
      (person.organizational_permissions || []).find(
        // @ts-ignore
        o => o.organization_id === orgId,
      );
    return dispatch({
      type: LOAD_PERSON_DETAILS,
      person,
      orgId,
      org: orgPermission && orgPermission.organization,
    });
  };
}

// @ts-ignore
export function savePersonNote(personId, notes, noteId, myId) {
  // @ts-ignore
  return dispatch => {
    if (!personId) {
      return Promise.reject(
        'Invalid Data from savePersonNote: no personId passed in',
      );
    }

    const bodyData = {
      data: {
        type: 'person_note',
        attributes: {
          content: notes || '',
        },
        relationships: {
          person: {
            data: {
              type: 'person',
              id: personId,
            },
          },
          user: {
            data: {
              type: 'user',
              id: myId,
            },
          },
        },
      },
    };

    if (!noteId) {
      return dispatch(callApi(REQUESTS.ADD_PERSON_NOTE, {}, bodyData));
    }
    return dispatch(callApi(REQUESTS.UPDATE_PERSON_NOTE, { noteId }, bodyData));
  };
}

// @ts-ignore
export function getPersonNote(personId, myId) {
  // @ts-ignore
  return dispatch => {
    const query = { person_id: personId, include: 'person_notes' };

    // @ts-ignore
    return dispatch(callApi(REQUESTS.GET_PERSON_NOTE, query)).then(results => {
      const person = results.find('person', personId);
      if (person && person.person_notes) {
        const notes = person.person_notes;
        // @ts-ignore
        return notes.find(element => {
          return element.user_id == myId;
        });
      }
      return Promise.reject('Person Not Found in getPersonNote');
    });
  };
}

// @ts-ignore
export function getPersonJourneyDetails(id) {
  // @ts-ignore
  return dispatch => {
    const query = {
      person_id: id,
      include:
        'pathway_progression_audits.old_pathway_stage,pathway_progression_audits.new_pathway_stage,interactions.comment,answer_sheets.answers,answer_sheets.survey.active_survey_elements.question',
    };
    // @ts-ignore
    return dispatch(callApi(REQUESTS.GET_PERSON_JOURNEY, query));
  };
}

// @ts-ignore
export function updatePersonAttributes(personId, personAttributes) {
  return {
    type: UPDATE_PERSON_ATTRIBUTES,
    updatedPersonAttributes: {
      id: personId,
      ...personAttributes,
    },
  };
}

// @ts-ignore
export function updatePerson(data) {
  const personInclude =
    'contact_assignments.person,email_addresses,phone_numbers,organizational_permissions.organization,reverse_contact_assignments,user';

  // @ts-ignore
  return async dispatch => {
    if (!(data && data.id)) {
      return dispatch({
        type: 'UPDATE_PERSON_FAIL',
        error: 'InvalidData',
        data,
      });
    }

    const updateData = { type: 'person' };
    let attributes;
    if (exists(data.firstName)) {
      attributes = { ...(attributes || {}), first_name: data.firstName };
    }
    if (exists(data.lastName)) {
      attributes = { ...(attributes || {}), last_name: data.lastName };
    }
    if (exists(data.gender)) {
      attributes = { ...(attributes || {}), gender: data.gender };
    }
    if (attributes) {
      // @ts-ignore
      updateData.attributes = attributes;
    }
    const bodyData = {
      data: updateData,
      ...(data.email || data.phone || data.orgPermission
        ? {
            included: [
              ...(data.email
                ? [
                    {
                      id: data.emailId,
                      type: 'email',
                      attributes: { email: data.email },
                    },
                  ]
                : []),
              ...(data.phone
                ? [
                    {
                      id: data.phoneId,
                      type: 'phone_number',
                      attributes: {
                        number: data.phone,
                      },
                    },
                  ]
                : []),
              ...(data.orgPermission && data.orgPermission.permission_id
                ? [
                    {
                      type: 'organizational_permission',
                      id: data.orgPermission.id,
                      attributes: {
                        permission_id: data.orgPermission.permission_id,
                      },
                    },
                  ]
                : []),
              ...(data.orgPermission && data.orgPermission.archive_date
                ? [
                    {
                      type: 'organizational_permission',
                      id: data.orgPermission.id,
                      attributes: {
                        archive_date: data.orgPermission.archive_date,
                      },
                    },
                  ]
                : []),
            ],
          }
        : {}),
    };
    const query = {
      personId: data.id,
      include: personInclude,
    };

    const results = await dispatch(
      callApi(REQUESTS.UPDATE_PERSON, query, bodyData),
    );
    const person = results.response;

    dispatch(
      updatePersonAttributes(data.id, {
        first_name: person.first_name,
        last_name: person.last_name,
        gender: person.gender,
        full_name: person.full_name,
        email_addresses: person.email_addresses,
        phone_numbers: person.phone_numbers,
        organizational_permissions: person.organizational_permissions,
      }),
    );

    return results;
  };
}

// @ts-ignore
export function makeAdmin(personId, orgPermissionId) {
  // @ts-ignore
  return async dispatch => {
    const results = await dispatch(
      updateOrgPermission(personId, orgPermissionId, ORG_PERMISSIONS.ADMIN),
    );
    dispatch(trackActionWithoutData(ACTIONS.MANAGE_MAKE_ADMIN));

    return results;
  };
}

// @ts-ignore
export function removeAsAdmin(personId, orgPermissionId) {
  // @ts-ignore
  return async dispatch => {
    const results = await dispatch(
      updateOrgPermission(personId, orgPermissionId, ORG_PERMISSIONS.USER),
    );
    dispatch(trackActionWithoutData(ACTIONS.MANAGE_REMOVE_ADMIN));

    return results;
  };
}

export function updateOrgPermission(
  // @ts-ignore
  personId,
  // @ts-ignore
  orgPermissionId,
  // @ts-ignore
  permissionLevel,
) {
  // @ts-ignore
  return dispatch => {
    const data = {
      id: personId,
      orgPermission: {
        id: orgPermissionId,
        permission_id: permissionLevel,
      },
    };
    return dispatch(updatePerson(data));
  };
}

// @ts-ignore
export function archiveOrgPermission(personId, orgPermissionId) {
  // @ts-ignore
  return async (dispatch, getState) => {
    const results = await dispatch(
      updatePerson({
        id: personId,
        orgPermission: {
          id: orgPermissionId,
          archive_date: new Date().toISOString(),
        },
      }),
    );

    const myId = getState().auth.person.id;
    dispatch(
      trackActionWithoutData(
        personId === myId
          ? ACTIONS.MANAGE_LEAVE_COMMUNITY
          : ACTIONS.MANAGE_REMOVE_MEMBER,
      ),
    );
    dispatch(getMyCommunities());

    return results;
  };
}

// @ts-ignore
export function updateFollowupStatus(person, orgPermissionId, status) {
  // @ts-ignore
  return async dispatch => {
    const data = {
      data: {
        type: 'person',
      },
      included: [
        {
          id: orgPermissionId,
          type: 'organizational_permission',
          attributes: {
            followup_status: status,
          },
        },
      ],
    };
    await dispatch(
      callApi(REQUESTS.UPDATE_PERSON, { personId: person.id }, data),
    );

    dispatch(trackActionWithoutData(ACTIONS.STATUS_CHANGED));

    return dispatch(
      updatePersonAttributes(person.id, {
        organizational_permissions: person.organizational_permissions.map(
          // @ts-ignore
          orgPermission =>
            orgPermission.id === orgPermissionId
              ? { ...orgPermission, followup_status: status }
              : orgPermission,
        ),
      }),
    );
  };
}

export function createContactAssignment(
  // @ts-ignore
  organizationId,
  // @ts-ignore
  personAssignedToId,
  // @ts-ignore
  personReceiverId,
) {
  // @ts-ignore
  return async dispatch => {
    const data = {
      included: [
        {
          type: 'contact_assignment',
          attributes: {
            assigned_to_id: personAssignedToId,
            organization_id: organizationId,
          },
        },
      ],
    };
    await dispatch(
      callApi(REQUESTS.UPDATE_PERSON, { personId: personReceiverId }, data),
    );
    dispatch(trackActionWithoutData(ACTIONS.ASSIGNED_TO_ME));
    return dispatch(getPersonDetails(personReceiverId, organizationId));
  };
}

// @ts-ignore
export function deleteContactAssignment(id, personId, personOrgId, note = '') {
  // @ts-ignore
  return async dispatch => {
    const data = {
      data: {
        type: 'contact_assignment',
        attributes: {
          unassignment_reason: note,
        },
      },
    };

    await dispatch(
      callApi(
        REQUESTS.DELETE_CONTACT_ASSIGNMENT,
        { contactAssignmentId: id },
        data,
      ),
    );

    dispatch(getMySteps());
    return dispatch({
      type: DELETE_PERSON,
      personId,
      personOrgId,
    });
  };
}

// @ts-ignore
export function navToPersonScreen(person, org, props = {}) {
  // @ts-ignore
  return (dispatch, getState) => {
    const organization = org ? org : {};
    const { auth, people, organizations } = getState();
    const orgId = organization.id;
    const personId = person.id;

    const selectorOrg =
      organizationSelector({ organizations }, { orgId }) || organization;
    //TODO Creating a new object every time will cause shallow comparisons to fail and lead to unnecessary re-rendering

    const selectorPerson =
      personSelector({ people }, { orgId, personId }) || person;

    const contactAssignment = contactAssignmentSelector(
      { auth },
      { person: selectorPerson, orgId },
    );
    const authPerson = auth.person;

    dispatch(
      navigatePush(
        getPersonScreenRoute(
          authPerson,
          selectorPerson,
          selectorOrg,
          contactAssignment,
        ),
        {
          ...props,
          person: selectorPerson,
          organization: selectorOrg,
        },
      ),
    );
  };
}

export function getPersonScreenRoute(
  // @ts-ignore
  mePerson,
  // @ts-ignore
  person,
  // @ts-ignore
  organization,
  // @ts-ignore
  contactAssignment,
) {
  const isMe = person.id === mePerson.id;

  const isMember = hasOrgPermissions(
    orgPermissionSelector(
      {},
      {
        person: person,
        organization,
      },
    ),
  );

  const isUserCreatedOrg = organization.user_created;
  const isGroups = mePerson.user.groups_feature;

  if (isMe) {
    if (isMember) {
      if (isGroups) {
        return IS_GROUPS_ME_COMMUNITY_PERSON_SCREEN;
      }

      return ME_COMMUNITY_PERSON_SCREEN;
    }

    return ME_PERSONAL_PERSON_SCREEN;
  }

  if (isMember) {
    if (isUserCreatedOrg) {
      return IS_USER_CREATED_MEMBER_PERSON_SCREEN;
    }
    if (isGroups) {
      return IS_GROUPS_MEMBER_PERSON_SCREEN;
    }

    return MEMBER_PERSON_SCREEN;
  }

  if (contactAssignment) {
    return CONTACT_PERSON_SCREEN;
  }

  return UNASSIGNED_PERSON_SCREEN;
}
