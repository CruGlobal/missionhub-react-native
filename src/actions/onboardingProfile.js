import uuidv4 from 'uuid/v4';
import { Crashlytics } from 'react-native-fabric';

import {
  COMPLETE_ONBOARDING,
  FIRST_NAME_CHANGED, LAST_NAME_CHANGED, PERSON_FIRST_NAME_CHANGED, PERSON_LAST_NAME_CHANGED,
  RESET_ONBOARDING_PERSON, UPDATE_ONBOARDING_PERSON,
} from '../constants';

import callApi, { REQUESTS } from './api';
import { updatePerson } from './person';


/*
A user is considered to have completed onboarding once they've:
1) selected a stage for themselves, and
2) selected a stage for a contact assignment
 */
export function completeOnboarding() {
  return { type: COMPLETE_ONBOARDING };
}

export function firstNameChanged(firstName) {
  return {
    type: FIRST_NAME_CHANGED,
    firstName: firstName,
  };
}

export function lastNameChanged(lastName) {
  return {
    type: LAST_NAME_CHANGED,
    lastName: lastName,
  };
}

export function createMyPerson(firstName, lastName) {
  const data = {
    code: uuidv4(),
    first_name: firstName,
    last_name: lastName,
  };

  return async(dispatch) => {
    const me = await dispatch(callApi(REQUESTS.CREATE_MY_PERSON, {}, data));
    Crashlytics.setUserIdentifier(`${me.person_id}`);
    return me;
  };
}

export function personFirstNameChanged(firstName) {
  return {
    type: PERSON_FIRST_NAME_CHANGED,
    personFirstName: firstName,
  };
}

export function personLastNameChanged(lastName) {
  return {
    type: PERSON_LAST_NAME_CHANGED,
    personLastName: lastName,
  };
}

export function createPerson(firstName, lastName, myId) {
  const data = {
    data: {
      type: 'person',
      attributes: {
        first_name: firstName,
        last_name: lastName,
      },
    },
    included: [
      {
        type: 'contact_assignment',
        attributes: {
          assigned_to_id: myId,
        },
      },
    ],
  };

  return (dispatch) => {
    return dispatch(callApi(REQUESTS.ADD_NEW_PERSON, {}, data));
  };
}

export function updateOnboardingPerson(data) {
  return (dispatch) => {
    return dispatch(updatePerson(data)).then((r) => {
      dispatch({ type: UPDATE_ONBOARDING_PERSON, results: r });
      return r;
    });
  };
}

export function resetPerson() {
  return { type: RESET_ONBOARDING_PERSON };
}
