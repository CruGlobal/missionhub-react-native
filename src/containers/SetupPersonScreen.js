import React, { Component } from 'react';
import { connect } from 'react-redux';
import { View, Keyboard, Image } from 'react-native';
import { translate } from 'react-i18next';
import styles from './SetupScreen/styles';
import { Button, Text, PlatformKeyboardAvoidingView, Flex } from '../components/common';
import Input from '../components/Input/index';
import { navigatePush } from '../actions/navigation';
import { personFirstNameChanged, personLastNameChanged } from '../actions/person';
import { createPerson } from '../actions/profile';
import { PERSON_STAGE_SCREEN } from './PersonStageScreen';

@translate()
class SetupPersonScreen extends Component {
  saveAndGoToGetStarted() {
    if (this.props.personFirstName) {
      Keyboard.dismiss();

      this.props.dispatch(createPerson(this.props.personFirstName, this.props.personLastName)).then(() => {
        this.props.dispatch(navigatePush(PERSON_STAGE_SCREEN, {
          section: 'onboarding',
          subsection: 'add person',
        }));
      });
    }
  }

  render() {
    const { t } = this.props;

    return (
      <PlatformKeyboardAvoidingView>
        <Flex value={1} />
        <Flex value={2} style={{ alignItems: 'center' }}>
          <Image source={require('../../assets/images/add_someone.png')} />
        </Flex>

        <Flex value={3} style={{ padding: 30 }}>
          <View>
            <Text style={styles.label}>{t('profileLabels.firstNameNickname')}</Text>
            <Input
              ref={(c) => this.personFirstName = c}
              onChangeText={(t) => this.props.dispatch(personFirstNameChanged(t))}
              value={this.props.personFirstName}
              autoFocus={true}
              returnKeyType="next"
              blurOnSubmit={false}
              onSubmitEditing={() => this.personLastName.focus()}
              placeholder={t('profileLabels.firstName')}
              placeholderTextColor="white"
            />
          </View>

          <View style={{ paddingTop: 30 }}>
            <Input
              ref={(c) => this.personLastName = c}
              onChangeText={(t) => this.props.dispatch(personLastNameChanged(t))}
              value={this.props.personLastName}
              returnKeyType="next"
              placeholder={t('profileLabels.lastNameOptional')}
              placeholderTextColor="white"
              blurOnSubmit={true}
            />
          </View>
        </Flex>

        <Flex value={1} align="stretch" justify="end">
          <Button
            type="secondary"
            onPress={() => this.saveAndGoToGetStarted()}
            text={t('next').toUpperCase()}
          />
        </Flex>
      </PlatformKeyboardAvoidingView>
    );
  }
}

const mapStateToProps = ({ personProfile }) => ({
  personFirstName: personProfile.personFirstName,
  personLastName: personProfile.personLastName,
});

export default connect(mapStateToProps)(SetupPersonScreen);
export const SETUP_PERSON_SCREEN = 'nav/SETUP_PERSON';
