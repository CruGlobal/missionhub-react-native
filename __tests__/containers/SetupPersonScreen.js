import 'react-native';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';

// Note: test renderer must be required after react-native.
import SetupPersonScreen from '../../src/containers/SetupPersonScreen';
import { testSnapshot, createMockStore } from '../../testUtils';
import * as profile from '../../src/actions/onboardingProfile';
jest.mock('../../src/actions/onboardingProfile');
import * as navigation from '../../src/actions/navigation';
import * as person from '../../src/actions/person';

const mockState = {
  personProfile: {
    personFirstName: '',
    personLastName: '',
  },
};

let store = createMockStore(mockState);

jest.mock('react-native-device-info');

it('renders correctly', () => {
  testSnapshot(
    <Provider store={store}>
      <SetupPersonScreen />
    </Provider>
  );
});


describe('setup person screen methods', () => {
  let component;
  Enzyme.configure({ adapter: new Adapter() });
  profile.createPerson = jest.fn();
  profile.updateOnboardingPerson = jest.fn();
  person.resetPerson = jest.fn();
  navigation.navigatePush = jest.fn();

  beforeEach(() => {
    store = createMockStore({
      ...mockState,
      personProfile: {
        ...mockState.personProfile,
        personFirstName: 'Test',
      },
    });
    const screen = shallow(
      <SetupPersonScreen />,
      { context: { store } },
    );

    component = screen.dive().dive().dive().instance();
  });

  it('navigates away', () => {
    component.navigate();

    expect(navigation.navigatePush).toHaveBeenCalledTimes(1);
  });

  it('saves and creates person', () => {
    component.saveAndGoToGetStarted();

    expect(profile.createPerson).toHaveBeenCalledTimes(1);
  });

  it('saves and updates person', () => {
    component.setState({ personId: 1 });
    component.saveAndGoToGetStarted();

    expect(profile.updateOnboardingPerson).toHaveBeenCalledTimes(1);
  });

  it('resets person on unmount', () => {
    component.componentWillUnmount();

    expect(profile.resetPerson).toHaveBeenCalledTimes(1);
  });
});
