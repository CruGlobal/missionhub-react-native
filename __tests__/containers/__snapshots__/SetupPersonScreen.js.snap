// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  keyboardVerticalOffset={0}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#007398",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-between",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {},
        Object {
          "flex": 1,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flex": 2,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../assets/images/add_someone.png",
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "padding": 30,
        },
        Object {
          "flex": 3,
        },
      ]
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#505256",
            },
            Object {
              "fontFamily": "SourceSansPro-Regular",
            },
            Object {
              "color": "#52C5DC",
              "fontSize": 12,
            },
          ]
        }
      >
        Profile_Label_FirstName
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCorrect={false}
        autoFocus={true}
        blurOnSubmit={false}
        multiline={false}
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="First Name"
        placeholderTextColor="white"
        returnKeyType="next"
        selectionColor="#007398"
        style={
          Array [
            Object {
              "backgroundColor": undefined,
              "borderBottomColor": "#52C5DC",
              "borderBottomWidth": 1,
              "color": "#ffffff",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 16,
              "letterSpacing": 0.25,
              "paddingVertical": 5,
            },
            Object {},
          ]
        }
        underlineColorAndroid={undefined}
        value=""
      />
    </View>
    <View
      style={
        Object {
          "paddingTop": 30,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCorrect={false}
        autoFocus={false}
        blurOnSubmit={true}
        multiline={false}
        onChangeText={[Function]}
        placeholder="Last Name (if you want)"
        placeholderTextColor="white"
        returnKeyType="next"
        selectionColor="#007398"
        style={
          Array [
            Object {
              "backgroundColor": undefined,
              "borderBottomColor": "#52C5DC",
              "borderBottomWidth": 1,
              "color": "#ffffff",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 16,
              "letterSpacing": 0.25,
              "paddingVertical": 5,
            },
            Object {},
          ]
        }
        underlineColorAndroid={undefined}
        value=""
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {},
        Object {
          "alignItems": "stretch",
          "flex": 1,
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits="button"
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#52C5DC",
              "height": 60,
              "justifyContent": "center",
            },
            null,
            Object {},
            null,
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#505256",
              },
              Object {
                "fontFamily": "SourceSansPro-Regular",
              },
              Array [
                Object {
                  "backgroundColor": "rgba(0,0,0,0)",
                  "color": "#ffffff",
                  "fontSize": 18,
                  "fontWeight": "500",
                  "letterSpacing": 1.5,
                  "textAlign": "center",
                },
                Object {},
              ],
            ]
          }
        >
          NEXT
        </Text>
      </View>
    </View>
  </View>
</View>
`;
